### ISSUE TITLE
BETA 5.11.29 unmarshal error for  all device types: UAP USW USG

### ISSUE THREAD

**Issue by vthawkeye** [davidnewhall/unifi-poller#47]
I'm trying to run unifi-poller and seem to be having some issues.  I am getting the following errors: 
[ERROR] json.Unmarshal(uap): json: cannot unmarshal object into Go value of type string
[ERROR] Enable Debug Logging to output the failed payload.
[ERROR] json.Unmarshal(usw): json: cannot unmarshal object into Go value of type string
[ERROR] Enable Debug Logging to output the failed payload.
[ERROR] json.Unmarshal(ugw): json: cannot unmarshal object into Go value of type string
[ERROR] Enable Debug Logging to output the failed payload.
[ERROR] json.Unmarshal(uap): json: cannot unmarshal object into Go value of type string
[ERROR] Enable Debug Logging to output the failed payload.

I have to AP-AC-lites, a switch and a USG.  I get no stats for the devices, though the client stats pull over fine.  I am running controller version 5.11.29.

**Comment by davidnewhall** [comment]
This error... is nearly useless. It's not indicating which field couldn't be unmarshalled. :( The only way I know how to debug this is to look at the payload. Please enable debug in your config and run this again. That will print out the payload and possibly allow me to find the problem. Attach that here (remove any sensitive info) and I'll investigate a bit more.

**Comment by davidnewhall** [comment]
Please also copy/paste the first few lines from the log. These should include the unifi-poller version and controller version.

**Comment by vthawkeye** [comment]
[log.txt](https://github.com/davidnewhall/unifi-poller/files/3299583/log.txt)
I attached the log output.  I tried first with I think version 1.2.3, then 1.3.0 and right now it's compiled from source, all with the same error.  I've scrubed all the IPs and mac addresses.  Let me know if that causes any problems.

**Comment by davidnewhall** [comment]
This is great. Give me a bit to parse through this and come up with some ideas. :D scrubbing IPs and MACs is perfectly fine!

**Comment by davidnewhall** [comment]
It's looking like Ubiquiti re-organized the structure of the output in the API JSON payload. This is pretty much sad times. 

I'm at a bit of a loss, but I do have questions for you that will help me decide how to proceed. 

First, I do not see controller 5.11.x on the download site: https://www.ui.com/download/unifi/ - where did you get it? (Maybe it's the cloudkey?) Have you tried this with a version of 5.10.x? Was it working? Last question: What version go did you compile the app with? I ask because we are getting different error values from the json module. Once I saw this error it was easy find the problem. 
Yours:` json: cannot unmarshal object into Go value of type string`
Mine:`json: cannot unmarshal object into Go struct field .ap of type string`

**Comment by davidnewhall** [comment]
Didn't mean to close, accidentally pressed the wrong button.  Looking forward to your reply. Thanks!

**Comment by vthawkeye** [comment]
The 5.11 series is the beta controller (for now).  I've not tried a 5.10 version.  I don't think downgrading from 5.11 to 5.10 is supported, but I'll see if it's possible.

I believe I used go 1.7.4 on debian.  The error you posted, was what I got from the pre-compiled version (I believe).  Go as a language and it's build system (dep) is new to me.

**Comment by davidnewhall** [comment]
Thank you for the update and information. Since this is a beta version, you have an opportunity to tell Ubiquiti it has a regression. Changing the API on a minor version should generally be a blocker. The change doesn't add new features, it seems to nest data inside another hash.

Your data, from a UAP looks like this:
```
  "stat": {
    "ap": {
      "site_id": "5c50dffe49464234c9dec0b9",
      "o": "ap",
      "oid": "00:00:00:00:00:00",
      "ap": "00:00:00:00:00:00",
```
Mine looks like this:
```
  "stat": {
    "site_id": "574e86994566ffb914a2683c",
    "o": "ap",
    "oid": "80:2a:a8:10:ae:87",
    "ap": "80:2a:a8:10:ae:87",
    "time": 1560200400000,
```

Notice in my output `ap` is a string, the MAC address. In your output, `ap` is a hash (this is normal json).  And inside that hash are all the values that once lived one level up. I'd be okay with them moving the values, but renaming a string to a hash breaks any downstream dependencies. As a developer, that feels like the kinds of things we try to catch in a beta test. I highly recommend reporting this in the beta program/forums. (I honestly don't know much about it). Maybe point them to this conversation when you submit your report so they can see what's changed and how it's a problem. The above json snippet is used by the code in [this file](https://github.com/golift/unifi/blob/master/uap_type.go#L169-L263), and it expects specific types (that's why you got an error). 

A bit more:
If you really are using go 1.7, I can't recommend enough to upgrade to a newer version. A lot of improvements have been made. Good job getting things to compile! I know it can be rough. I'm looking into why the error seems to be missing output (from the pre-built binaries).

`dep` is a dependency manager. When you ran `dep ensure` it created a `vendor` folder that comes from the `Gopkg` files in the repo. I created those files by typing `dep init ; dep ensure -update`. This allows you to download the same dependencies I built and tested with. Pretty snazzy!

I'm sorry I can't be more help. I'm reluctant to jump in and make changes to support a beta version of the software. It's very possible these things can change before general release. The changes I'd have to implement aren't very straightforward either, or believe me I'd hook you up with a fix. Good luck!

**Comment by vthawkeye** [comment]
Thanks for taking a look.  I'll file a bug report and see what they say.  It's a little more than a Minor version though, the 5.11 series is in testing now, but will be become the stable release at some point with it's new features.  If I hear anything back about it, I'll let you know.

As for go, go 1.7 is what debian installs if you just say "apt install golang".  I now also installed 1.11 and used it to compile.  I've made some local changes to the uap/usw/usg_type/influx files and got stats coming in, though the changes would break against 5.10 :(

**Comment by davidnewhall** [comment]
I'm not using the word `minor` to indicate a small or insignificant change. Minor version is literally what it's called, and we call it that because it does not introduce backward-incompatible changes.

5 is the major version.
10 is the minor version.
29 is the patch level
See: https://semver.org/

And yes, it's entirely possible to make a few small changes to the two `uap_*.go` files in the unifi library to make this work for you. It just wont work for 5.10 or earlier after doing so. That's probably fine, but it becomes a problem when more updates are released and you have to keep patching to stay current.

I'm happy to leave this Issue open until 5.11 goes general, that way others may quickly find this if they're also beta testers. I'll also see your updates that way.

**Comment by tfs** [comment]
I took information from this thread and opened a issue on the beta community forum.

**Comment by davidnewhall** [comment]
In the hopes of supporting this new json payload format, it would be helpful if those running the beta software uploaded client and device payloads so I can incorporate their format. The data vthawkeye provided is actually a really good start, and will allow me to begin working on this. I've found that payloads can differentiations from controller to controller even if they run the same version. Some devices provide different options, etc.

tl;dr: I'm working on this.

**Comment by Waffull** [comment]
Since I'm on controller v5.11.31, unifi-poller isn't pulling data from both of my sites.  Therefore the json files will be somewhat incomplete.  I've scrubbed them and hopefully didn't leave anything sensitive in.

[Archive.zip](https://github.com/davidnewhall/unifi-poller/files/3366105/Archive.zip)

Hope these help.

**Comment by davidnewhall** [comment]
Thank you all for the provided data! (some was emailed)

I think I have a solution. Is anyone able to compile from source? Wanna give beta software a try? :D irony.

**Comment by davidnewhall** [comment]
@Waffull - When you say it's not pulling data from both sites.. can you be more descriptive? The client json appears to unaffected, so I'd expect you to see accurate client counts for both sites. Is it just not storing any data for the second site? Is it only storing data for the `default` site? Do this for me: enable debug mode and when the app starts (and every 30 seconds) it prints a debug line with all the sites the library found. It looks like this:
```
2019/07/07 14:40:14.028218 unifi.go:141: [DEBUG] Found 1 site(s): default
```
Does this line list more than one site (for anyone with controller 5.11) ? I'm going to edit this comment in a moment with a debug comment to run.....

**Comment by Waffull** [comment]
It only shows one site... which is calls default.  It's ignoring a second
site for some reason.  I posted the debug info in the UI community
https://community.ui.com/questions/Unifi-Poller-Store-Unifi-Controller-Metrics-in-InfluxDB-without-SNMP/58a0ea34-d2b3-41cd-93bb-d95d3896d1a1?page=2
Sorry, I don't have a copy of it on me.

Can do more after I get home this evening if you need.

**Comment by davidnewhall** [comment]
Yeah I saw that debug info, I don't think I need more of the log output at this point. Thanks! I'm convinced something changed in the sites api. The output of the debug command above will help a lot!

**Comment by Waffull** [comment]
Well now I feel like a complete idiot...  Figured out why only the default site was showing up.... I hadn't created the influxdb admin in the 2nd site.  Might be worth putting a note in the wiki that the admin has to be added to each site separately.

Anyway, here's the debug file, in case it can be of any help at this point.
[sites.txt](https://github.com/davidnewhall/unifi-poller/files/3366206/sites.txt)

**Comment by davidnewhall** [comment]
Sweet! You're now able to collect at least the client data from each site, yes? This is great feedback! No need to feel any way about it, this is all a learning experience. A couple months ago I had no idea what a site was!

Looking at your sites json output, I would expect the app to at least print out both site names on startup. I'm actually surprised the auth info is per site. I guess that gates access to even *seeing* the site? That's interesting and makes the log output (lacking site 2) make sense.

This is good info. The sites output looks the same as 5.10, so I think it's fine.

I'll push a new release of the library shortly and then update `master` branch on unifi-poller to reference it. That will trigger a new docker build at least, but not new binaries for other platforms. Once we get the docker tested (I think @tfs can do that for us), I'll release a new version of unifi-poller - solving this problem. This task will close, but I'll re-open it until we have confirmation.

**Comment by davidnewhall** [comment]
Please test and report.
```shell
docker pull golift/unifi-poller:latest
```
^ Multi architecture. Works on arm or x86.

**Comment by vthawkeye** [comment]
I get an error from both latest and 5_11:

docker: Error response from daemon: OCI runtime create failed: container_linux.go:345: starting container process caused "exec: \"/${BINARY}\": stat /${BINARY}: no such file or directory": unknown.

**Comment by davidnewhall** [comment]
try 5_11_support

**Comment by Waffull** [comment]
Yes, I can see client data for both sites :)  No USG or USW data and no data in UAP graphs, but I suspect the payload changes will fix that.

I'll check out the wiki and add anything that's necessary.  Thanks.

**Comment by vthawkeye** [comment]
Running the docker version with no issues so far.  Updated dashboards and only notice the links to the other dashboards do not show all the other dashboards on the clients and UAP dashboards.  They only show the ones tagged with unifi-poller and not unifi.  Also, those two have "UniFi-Poller" and the others have "unifi-poller" as the first part of their names.

Hope that helps,

**Comment by davidnewhall** [comment]
Awesome! What you described is sorta normal. I'll look again to make sure all my tags are correct. I'll also update the other 3 dashboards so the names match.

**Comment by davidnewhall** [comment]
The new release is cut and building in Travis-CI and Docker Cloud now. https://github.com/davidnewhall/unifi-poller/releases/tag/v1.4.2

The Docker Cloud build is new, so very hopeful this works! Travis CI will upload the new packages to the release on github soon.

**Comment by Waffull** [comment]
So all the dashboards are working except the USG.  Seems unifi-poller isn't seeing my 2 USGs.
Here's the non-debug log.  I'll send the debug if that will help.
Thanks.

> mschauber@minipc:~$ tail -f -n100  /var/log/syslog /var/log/messages | grep unifi-poller
> tail: cannot open '/var/log/messages' for reading: No such file or directoryJul  8 03:09:25 minipc unifi-poller[13467]: 2019/07/08 03:09:25 [ERROR] json.Unmarshal(ugw): json: cannot unmarshal bool into Go struct field .up of type string
> Jul  8 03:09:25 minipc unifi-poller[13467]: 2019/07/08 03:09:25 [ERROR] Enable Debug Logging to output the failed payload.
> Jul  8 03:09:25 minipc unifi-poller[13467]: 2019/07/08 03:09:25 [ERROR] json.Unmarshal(ugw): json: cannot unmarshal bool into Go struct field .up of type string
> Jul  8 03:09:25 minipc unifi-poller[13467]: 2019/07/08 03:09:25 [ERROR] Enable Debug Logging to output the failed payload.
> Jul  8 03:09:25 minipc unifi-poller[13467]: 2019/07/08 03:09:25 [INFO] UniFi Measurements Recorded. Sites: 2, Clients: 24, Wireless APs: 4, Gateways: 0, Switches: 3, Points: 187, Fields: 6319
> Jul  8 03:09:55 minipc unifi-poller[13467]: 2019/07/08 03:09:55 [ERROR] json.Unmarshal(ugw): json: cannot unmarshal bool into Go struct field .up of type string
> Jul  8 03:09:55 minipc unifi-poller[13467]: 2019/07/08 03:09:55 [ERROR] Enable Debug Logging to output the failed payload.
> Jul  8 03:09:55 minipc unifi-poller[13467]: 2019/07/08 03:09:55 [ERROR] json.Unmarshal(ugw): json: cannot unmarshal bool into Go struct field .up of type string
> Jul  8 03:09:55 minipc unifi-poller[13467]: 2019/07/08 03:09:55 [ERROR] Enable Debug Logging to output the failed payload.
> Jul  8 03:09:55 minipc unifi-poller[13467]: 2019/07/08 03:09:55 [INFO] UniFi Measurements Recorded. Sites: 2, Clients: 24, Wireless APs: 4, Gateways: 0, Switches: 3, Points: 187, Fields: 6319

**Comment by davidnewhall** [comment]
oh shoot you're getting an unmarshal error on something else. lemme look...

**Comment by davidnewhall** [comment]
it's actually a really easy fix, but man, what a way to kill me happy release mood. haha

**Comment by Waffull** [comment]
I'm not setup to compile right now.  I can give it a go in the AM.  I have to be up in 5.5 hours for work, so don't want to get into a process that may take some time.  If you can compile, I'll give it a test, yes, amd64 deb.  If not, I should have some time tomorrow before noon.

Sorry about bursting your happy release mood.  The good news is that most other things are working and I'm really impressed with all the data and your dashboards.

**Comment by davidnewhall** [comment]
The fix is committed to master. Here's a package built from master. Lemme know!

**Comment by Waffull** [comment]
Sorry, had to take the dog out.  
USGs now being seem :)  A few errors on the dashboard... the triangle errors say 'multi series error'

Off to sleep.  Thanks for everything.  I'll get back to you for any replies in the AM.

**Comment by davidnewhall** [comment]
Sounds like you might be using the old UAP dashboard? Make sure you have the latest dashboards.

**Comment by davidnewhall** [comment]
Here try this. 
[UniFi-Poller_ USG Insights-1562557956475.json.gz](https://github.com/davidnewhall/unifi-poller/files/3366626/UniFi-Poller_.USG.Insights-1562557956475.json.gz)

**Comment by Waffull** [comment]
Have all the newest dashboards (the ones you updated about an hour ago.)
The errors on the USG dashboard are because I have multiple USGs, it's
trying to populate those fields with data from both, and that can't be
done.  If I change to one of the sites, they work fine.

Must go to sleep.  Goodnight.

**Comment by davidnewhall** [comment]
Good night!

**Comment by davidnewhall** [comment]
I uploaded a new revision of the USG dashboard to Grafana. It's slightly different than the one I uploaded on my previous comment.

**Comment by tfs** [comment]
Ok, starting to test this. Shall I still use 5_11_support? Or is :latest in sync, too? (Version number on latest is higher than on the 5_11 one...) Poller runs without error messages.

[INFO] UniFi Measurements Recorded. Sites: 2, Clients: 28, Wireless APs: 3, Gateways: 1, Switches: 3, Points: 95, Fields: 3963

**Comment by davidnewhall** [comment]
`latest` **right now** should be OK. `stable` may work too, but Waffull ran into a bug that's now fixed (and will be pushed into `stable` soon). What's going on with grafana? Feel free to post or email a screenshot.

**Comment by tfs** [comment]
Hi,

My bad. Had a configuration problem with my Database from some of my experiments.

Now its working fine - using :latest and the new dashboards.
Wonderful!

As for the Synology mentions: I’m running Plex on the 918+ - and it uses the hardware for decoding just fine. Works even for 4k streams.

Now off to work :) Have a fine day!!

Thanks, Thomas.

**Comment by Waffull** [comment]
The updated USG dashboard is great.
I'm having some trouble compiling an armhf build.  Having some sort of GOROOT & GOPATH problem.  Working on that now.  Will let you know if I need some assistance.

**Comment by davidnewhall** [comment]
Glad everyone's working! Feel free to crack open a new issue and I'm happy to help! I'm pretty good with go.

**Comment by rottison** [comment]
hello I am new to unifi and thought this would be a great way to see whats happening on my network I am using this in my home network. I installed unifi-poller_1.4.2-133_armhf.deb manually and it seemed to work fine I have edited the up file in etc/unifi-poller it is on a Pi3 B+ with the unifi controller software
it is putting out Jul 10 12:09:43 controller unifi-poller[3102]: 2019/07/10 12:09:43 [INFO] UniFi Measurements Recorded. Sites: 1, Clients: 18, Wireless APs: 2, Gateways: 1, Switches: 1, Points: 51, Fields: 2186 but no data is making it to Grafana dashboard I made the new data source and used the button on the bottom for testing it and it turns green says its working in grafana i have the dashboards poingted to the new data source named it unifi and using google found some commands to see the data in the influx database. I have unfi-poller and unifi controller on the pi and influx and grafana run on a UnRaid server as Dockers I have other databases in influx that work for monitoring UnRaid systeam

**Comment by davidnewhall** [comment]
@rottison - Can you open up a new Issue for this?  This issue is close and seems unrelated to the problem you're describing. I'm happy to help, and if you open a new issue it'll be easier for me to find.

Please also make sure you've downloaded the [latest dashboards from Grafana](https://grafana.com/dashboards?search=unifi-poller).

If you're seeing the data in Influx (using SELECT statements) then unifi-poller is working (the log output you provided looks healthy). I think we're good there, so the problem is either in the Grafana data source or the dashboards (using the wrong data source maybe?).

Try making a new dashboard/graph and see if you can graph any of the data in the `unifi` data source.
